# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1426437665 14400
#      Sun Mar 15 12:41:05 2015 -0400
# Node ID 6e5365ee200a624f5b4f9540053794e6b6e42548
# Parent  4910b70d73284dd4388aec081ee4387b6257d57d
ginput: Fix for clicks outside axes and within multiple axes (bug #44426)

* ginput.m: Use the figure "windowbuttondownfcn" callback instead of the
axes "buttondownfcn" callback.  Get mouse click coordinates from gca.

diff --git a/scripts/plot/util/ginput.m b/scripts/plot/util/ginput.m
--- a/scripts/plot/util/ginput.m
+++ b/scripts/plot/util/ginput.m
@@ -59,22 +59,22 @@ function varargout = ginput (n = -1)
       [varargout{:}] = feval (toolkit_fcn, fig, n);
     endif
     return
   endif
 
   x = y = button = [];
   ginput_accumulator (0, 0, 0, 0);  # initialize accumulator
 
-  orig_buttondownfcn = get (fig, "buttondownfcn");
+  orig_windowbuttondownfcn = get (fig, "windowbuttondownfcn");
   orig_ginput_keypressfcn = get (fig, "keypressfcn");
 
   unwind_protect
 
-    set (ax, "buttondownfcn", @ginput_buttondownfcn);
+    set (fig, "windowbuttondownfcn", @ginput_windowbuttondownfcn);
     set (fig, "keypressfcn", @ginput_keypressfcn);
 
     do
       if (strcmp (toolkit, "fltk"))
         __fltk_check__ ();
       endif
 
       ## Release CPU.
@@ -86,17 +86,17 @@ function varargout = ginput (n = -1)
     if (n0 > n)
       ## More clicks than requested due to double-click or too fast clicking
       x = x(1:n);
       y = y(1:n);
       button = button(1:n);
     endif
 
   unwind_protect_cleanup
-    set (ax, "buttondownfcn", orig_buttondownfcn);
+    set (fig, "windowbuttondownfcn", orig_windowbuttondownfcn);
     set (fig, "keypressfcn", orig_ginput_keypressfcn);
   end_unwind_protect
 
   varargout = {x, y, button};
 
 endfunction
 
 function [x, y, n, button] = ginput_accumulator (mode, xn, yn, btn)
@@ -114,18 +114,18 @@ function [x, y, n, button] = ginput_accu
     n += 1;
   elseif (mode == 2)
     ## The end due to Enter.
     n = -1;
  endif
 
 endfunction
 
-function ginput_buttondownfcn (src, button)
-  point = get (src, "currentpoint");
+function ginput_windowbuttondownfcn (src, button)
+  point = get (gca (), "currentpoint");
   ginput_accumulator (1, point(1,1), point(1,2), button);
 endfunction
 
 function ginput_keypressfcn (src, evt)
   point = get (gca (), "currentpoint");
   key = evt.Key;
   if (key == "return")
     ## Enter key stops ginput.
